### PR TITLE
update xstate to v5

### PR DESCRIPTION
**Pull request by sabrine33**

### CODE REVIEW

**Review by khelwood — COMMENTED** [review]
Well done getting through all this. I trust that it all makes sense; I've just put a few questions to help my own understanding.

File: src/lib/machines/extraction/extractionMachine.ts [hunk 1]
```diff
@@ -96,53 +124,41 @@ export const extractionMachine = createMachine<ExtractionContext, ExtractionEven
...
-        if (e.type !== 'UPDATE_EQUIPMENT_ID') return;
-        ctx.equipmentId = e.equipmentId;
+      assignEquipmentId: assign(({ context, event }) => {
+        if (event.type !== 'UPDATE_EQUIPMENT_ID') return context;
+        context.equipmentId = event.equipmentId;
+        return context;
       })
-    },
-
-    services: {
```
[the remark this answers is not in the document]
Comment (sabrine33) [reply src/lib/machines/extraction/extractionMachine.ts:131]: In XState v5, the services section was removed, and all tasks that were previously defined as services should now be handled as actions.

File: src/lib/machines/analysis/analysisMachine.ts [hunk 2]
```diff
@@ -1,166 +0,0 @@
-import { ExtractResult, PassFail, RecordRnaAnalysisMutation, RnaAnalysisRequest } from '../../../types/sdk';
```
[the remark this answers is not in the document]
Comment (sabrine33) [reply src/lib/machines/analysis/analysisMachine.ts:1]: I deleted it as it is not used anywhere.

File: src/lib/factories/labwareTypeFactory.ts [hunk 3]
```diff
@@ -3,58 +3,63 @@ import { LabwareType } from '../../types/sdk';
...
-export default labwareTypeFactory;
-
 export const labwareTypes: Record<LabwareTypeName, Factory<LabwareType>> = {
   [LabwareTypeName.TUBE]: labwareTypeFactory.params({
     __typename: 'LabwareType',
     name: LabwareTypeName.TUBE,
     numRows: 1,
-    numColumns: 1
+    numColumns: 1,
+    labelType: labelTypeFactory.build({ name: `${LabwareTypeName.TUBE} Label` })
```
[the remark this answers is not in the document]
Comment (sabrine33) [reply src/lib/factories/labwareTypeFactory.ts:20]: The factory files are only used for the mocking server. The previous way was affecting a sequence number to the label type name, when filtering by the labware label type name, it was leading to some cypress tests fails.

File: src/components/workAllocation/workRow.machine.ts [hunk 4]
```diff
@@ -149,7 +158,16 @@ export default function createWorkRowMachine({ workWithComment }: CreateWorkRowM
         },
         updating: {
           invoke: {
-            src: 'updateWorkStatus',
```
[the remark this answers is not in the document]
Comment (sabrine33) [reply src/components/workAllocation/workRow.machine.ts:152]: the fromPromise is to run async actions, this is to replace the services

File: src/components/workAllocation/workAllocation.machine.ts [hunk 5]
```diff
@@ -281,84 +329,56 @@ export default function createWorkAllocationMachine({ urlParams }: CreateWorkAll
...
+            draft.projects.push(event.project);
+          });
         }),
-        addNewlyCreatedCostCode: assign((ctx, event) => {
-          if (event.type !== 'ADD_NEWLY_CREATED_COST_CODE') return;
-          ctx.costCodes.push(event.costCode);
-        })
-      },
-
-      services: {
```
[the remark this answers is not in the document]
Comment (sabrine33) [reply src/components/workAllocation/workAllocation.machine.ts:326]: The services section is been removed from the machine definition, async actions should be defined within the actions section using fromPromise

File: src/components/slotMapper/slotMapper.machine.ts [hunk 6]
```diff
@@ -1,246 +1,277 @@
...
         }
       }
+      return { ...context, failedSlots };
     }),
-    assignPassFailError: assign((ctx, e) => {
-      if (e.type !== 'error.platform.passFailsSlots') return;
-      ctx.errors.set(e.barcode, e.error);
-    })
-  },
-  services: {
```
[the remark this answers is not in the document]
Comment (sabrine33) [reply src/components/slotMapper/slotMapper.machine.ts:227]: The services section is been removed from the machine definition, async actions should be defined within the actions section using fromPromise

File: src/components/planning/labwarePlan.machine.ts [hunk 7]
```diff
@@ -161,58 +198,40 @@ export const createLabwarePlanMachine = (initialLayoutPlan: LayoutPlan) =>
...
 
       guards: {
-        isVisiumLP: (ctx) => ctx.layoutPlan.destinationLabware.labwareType.name === LabwareTypeName.VISIUM_LP,
+        isVisiumLP: ({ context }) =>
+          context.layoutPlan.destinationLabware.labwareType.name === LabwareTypeName.VISIUM_LP,
 
-        isLayoutValid: (ctx) => ctx.layoutPlan.plannedActions.size > 0
-      },
-
-      services: {
```
[the remark this answers is not in the document]
Comment (sabrine33) [reply src/components/planning/labwarePlan.machine.ts:193]: The services section is been removed from the machine definition, async actions should be defined within the actions section using fromPromise

File: src/components/labwareScanner/LabwareScanner.tsx [hunk 8]
```diff
@@ -89,50 +89,56 @@ export default function LabwareScanner({
...
+
+        if (currentLabwareLength > previousLabwareLength) {
+          onAdd?.(labwares[labwares.length - 1]);
+        }
+
+        if (currentLabwareLength < previousLabwareLength) {
+          observer.context.removedLabware &&
+            onRemove?.(observer.context.removedLabware.labware, observer.context.removedLabware.index);
+        }
+      }
```
Comment (seenanair) [inline src/components/labwareScanner/LabwareScanner.tsx:133]: Just to understand, what this change means?

File: src/pages/StainingQC.tsx [hunk 9]
```diff
@@ -42,24 +43,24 @@ export default function StainingQC({ info }: StainingQCProps) {
   const stanCore = useContext(StanCoreContext);
 
   const formMachine = React.useMemo(() => {
-    return createFormMachine<ResultRequest, RecordStainResultMutation>().withConfig({
-      services: {
-        submitForm: (ctx, e) => {
-          if (e.type !== 'SUBMIT_FORM') return Promise.reject();
+    return createFormMachine<ResultRequest, RecordStainResultMutation>().provide({
+      actors: {
+        submitForm: fromPromise(({ input }) => {
```
Comment (seenanair) [inline src/pages/StainingQC.tsx:48]: I can see lot of these changes. I hope this is the new xstate way it should be used

**Review by seenanair — APPROVED** [review]
Great piece of work.Well done - to all the effort you put in. 👍

**Comment by sabrine33** [comment]
merged to devel

**Pull request merged**